wire-ios/Wire-iOS/Sources/UserInterface/Calling/Analytics/CallEndedAnalyticsController.swift

```diff
@@ -0,0 +1,288 @@
+        self.notificationCenter = notificationCenter
+        self.analyticsEventTracker = analyticsEventTracker
+        self.logger = logger
+        self.currentDateProvider = currentDateProvider
+
+        setupObservations()
+    }
+
+    deinit {
+        let notificationCenter = notificationCenter
+        if let toggleVideoObservationToken {
+            notificationCenter.removeObserver(toggleVideoObservationToken)
+        }
+    }
+
+    private func setupObservations() {
+
+        callStateObservationToken = CallCenter.addCallStateObserver(
+            observer: self,
+            contextProvider: contextProvider
+        )
+
+        toggleVideoObservationToken = notificationCenter.addObserver(
+            forName: WireCallCenterV3.didToggleVideoNotification, object: .none, queue: .none
+        ) { [weak self] notification in
+            self?.handleToggleVideoNotification(notification)
+        }
+    }
+
+    private func handleIncomingCall(
+        _ conversation: ZMConversation,
+        _ isVideoCall: Bool
+    ) {
+        if let eventInfo {
+            logger.error("handleIncomingCall: expected eventInfo to be nil, but is: \(eventInfo)")
+        }
+
+        eventInfo = .init(
+            callStart: currentDateProvider.now,
+            conversationID: conversation.avsIdentifier,
+            conversationType: conversation.conversationType == .group ? .group : .oneOnOne,
+            isVideoCall: isVideoCall
+        )
+    }
+
+    private func handleOutgoingCall(
+        _ conversation: ZMConversation,
+        _ isVideoCall: Bool
+    ) {
+        if let eventInfo {
+            logger.error("handleOutgoingCall: expected eventInfo to be nil, but is: \(eventInfo)")
+        }
+
+        eventInfo = .init(
+            callDirection: .outgoing,
+            callStart: currentDateProvider.now,
+            conversationType: conversation.conversationType == .group ? .group : .oneOnOne,
+            isVideoCall: isVideoCall
+        )
+    }
+
+    private func handleCallEstablished(_ conversation: ZMConversation) {
+        if eventInfo == nil {
+            logger.error("handleCallEstablished: expected eventInfo to be non-nil")
+            eventInfo = .init(callStart: currentDateProvider.now)
+        }
+
+        callParticipantObsererToken = CallCenter.addCallParticipantObserver(
+            observer: self,
+            for: conversation,
+            contextProvider: contextProvider
+        )
+    }
+
+    private func handleToggleVideoNotification(_ notification: Notification) {
+        guard
+            let conversationID = notification.userInfo?[CallCenter.conversationIDUserInfoKey] as? AVSIdentifier,
+            conversationID == eventInfo?.conversationID
+        else { return }
+
+        eventInfo?.hasAVSwitchToggled = true
+
+        // unsubscribe
+        toggleVideoObservationToken.map { notificationCenter.removeObserver($0) }
+        toggleVideoObservationToken = nil
```

Comment:
This is because we only care about a single toggle right? If so, perhaps a comment would be useful for others.